src/Middleware/RateLimiting/src/RateLimitingMiddleware.cs

```diff
@@ -0,0 +1,82 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Threading.RateLimiting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.RateLimiting;
+
+/// <summary>
+/// Limits the rate of requests allowed in the application, based on limits set by a user-provided <see cref="PartitionedRateLimiter{TResource}"/>.
+/// </summary>
+internal sealed partial class RateLimitingMiddleware
+{
+    private readonly RequestDelegate _next;
+    private readonly Func<HttpContext, RateLimitLease, Task> _onRejected;
+    private readonly ILogger _logger;
+    private readonly PartitionedRateLimiter<HttpContext> _limiter;
+    private readonly int _rejectionStatusCode;
+
+    /// <summary>
+    /// Creates a new <see cref="RateLimitingMiddleware"/>.
+    /// </summary>
+    /// <param name="next">The <see cref="RequestDelegate"/> representing the next middleware in the pipeline.</param>
+    /// <param name="logger">The <see cref="ILogger"/> used for logging.</param>
+    /// <param name="options">The options for the middleware.</param>
+    public RateLimitingMiddleware(RequestDelegate next, ILogger<RateLimitingMiddleware> logger, IOptions<RateLimiterOptions> options)
+    {
+        _next = next ?? throw new ArgumentNullException(nameof(next));
+
+        if (logger == null)
+        {
+            throw new ArgumentNullException(nameof(logger));
+        }
+
+        _logger = logger;
+        _limiter = options.Value.Limiter;
+        _onRejected = options.Value.OnRejected;
+        _rejectionStatusCode = options.Value.DefaultRejectionStatusCode;
+    }
+
+    // TODO - EventSource?
+    /// <summary>
+    /// Invokes the logic of the middleware.
+    /// </summary>
+    /// <param name="context">The <see cref="HttpContext"/>.</param>
+    /// <returns>A <see cref="Task"/> that completes when the request leaves.</returns>
+    public async Task Invoke(HttpContext context)
+    {
+        using var lease = await TryAcquireAsync(context);
+        if (lease.IsAcquired)
+        {
+            await _next(context);
+        }
+        else
+        {
+            RateLimiterLog.RequestRejectedLimitsExceeded(_logger);
```

Comment:
Should we include `context.TraceIdentifier` in this log? It might be better to use or `Activity.Current?.Id ?? context.TraceIdentifier`, but then we'd want to gate the async local lookup in `Activity.Current` to only when debug logs are enabled. I think `context.TraceIdentifier` is better than nothing and good enough for now.

Reply:
Never mind. This can be left for later. `context.TraceIdentifier` is also lazily initialized so we'd want to gate that too. For now people can use log scopes if they really want to see which requests were rejected. Or log in a custom handler.